### PR TITLE
Convert `rz_str_new` to `rz_str_dup`

### PR DESCRIPTION
**Pull request by HN026**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Partially addresses #1176
...

### CODE REVIEW

File: librz/include/rz_util/rz_str.h [hunk 1]
```diff
@@ -194,7 +193,7 @@ static inline const char *rz_str_get_null(const char *str) {
 	return str ? str : "(null)";
 }
 RZ_API char *rz_str_ndup(RZ_NULLABLE const char *ptr, int len);
-RZ_API char *rz_str_dup(char *ptr, const char *string);
+RZ_API char *rz_str_dup(const char *str);
```
Comment (wargio) [inline librz/include/rz_util/rz_str.h]: add RZ_NULLABLE for `str` and `RZ_OWN` for the result.

Comment (wargio) [reply librz/include/rz_util/rz_str.h]: same in the .c function

File: librz/util/str.c [hunk 2]
```diff
@@ -1047,10 +1041,8 @@ RZ_API char *rz_str_ndup(RZ_NULLABLE const char *ptr, int len) {
 }
 
 // TODO: deprecate?
```
Comment (wargio) [inline librz/util/str.c]: change this comment and add a doxygen comment on what the function do

**Review by wargio — CHANGES_REQUESTED** [review]
nice changes!

File: librz/core/tui/config.c [hunk 3]
```diff
@@ -36,7 +36,8 @@ static void config_visual_hit(RzCore *core, const char *name, int editor) {
 		// XXX: must use config_set () to run callbacks!
 		if (editor) {
 			char *buf = rz_core_editor(core, NULL, node->value);
-			node->value = rz_str_dup(node->value, buf);
+			free(node->value);
+			node->value = rz_str_dup(buf);
```
Comment (wargio) [inline librz/core/tui/config.c]: probably you should dup before freeing.
```c
char *tmp = rz_str_dup(buf);
free(node->value);
node->value = tmp;
```
And this probably needs to be done for all the old invocations of `rz_str_dup`

Comment (HN026) [reply librz/core/tui/config.c]: Sureee

File: librz/include/rz_util/rz_str.h (hunk 1 above)
Comment (HN026) [reply librz/include/rz_util/rz_str.h]: Okayy

File: librz/core/tui/panels.c [hunk 4]
```diff
@@ -2846,10 +2846,12 @@ void __replace_cmd(RzCore *core, const char *title, const char *cmd) {
...
 	__free_panel_model(cur);
 	cur->model = RZ_NEW0(RzPanelModel);
+	char *tmp1 = rz_str_dup(title);
 	free(cur->model->title);
-	cur->model->title = rz_str_dup(title);
+	cur->model->title = tmp1;
+	char *tmp2 = rz_str_dup(cmd);
 	free(cur->model->cmd);
-	cur->model->cmd = rz_str_dup(cmd);
+	cur->model->cmd = tmp2;
```
Comment (wargio) [inline librz/core/tui/panels.c]: You can define `tmp` once and just reused it.
```c
	char *tmp = rz_str_dup(title);
	free(cur->model->title);
	cur->model->title = tmp;
	tmp = rz_str_dup(cmd);
	free(cur->model->cmd);
	cur->model->cmd = tmp;
```

File: librz/core/tui/panels.c [hunk 5]
```diff
@@ -3010,25 +3013,32 @@ void __init_panel_param(RzCore *core, RzPanel *p, const char *title, const char
...
+		m->cmd = tmp2;
 	} else {
+		char *tmp1 = rz_str_dup("");
 		free(m->title);
-		m->title = rz_str_dup("");
+		m->title = tmp1;
+		char *tmp2 = rz_str_dup("");
 		free(m->cmd);
-		m->cmd = rz_str_dup("");
+		m->cmd = tmp2;
```
Comment (wargio) [inline librz/core/tui/panels.c]: Same here. define it once within the if and then reuse it.

File: librz/include/rz_util/rz_str.h [hunk 6]
```diff
@@ -193,7 +193,7 @@ static inline const char *rz_str_get_null(const char *str) {
 	return str ? str : "(null)";
 }
 RZ_API char *rz_str_ndup(RZ_NULLABLE const char *ptr, int len);
-RZ_API char *rz_str_dup(const char *str);
+RZ_API char *rz_str_dup(RZ_NULLABLE const char *str);
```
Comment (wargio) [inline librz/include/rz_util/rz_str.h]:
```suggestion
RZ_API RZ_OWN char *rz_str_dup(RZ_NULLABLE const char *str);
```

File: librz/util/str.c [hunk 7]
```diff
@@ -1040,8 +1040,16 @@ RZ_API char *rz_str_ndup(RZ_NULLABLE const char *ptr, int len) {
...
+/**
+ * @brief Duplicates a string.
+ *
+ * This function duplicates the given string. If the input string is NULL,
+ * the function will return NULL.
+ *
+ * @param str The string to duplicate. Can be NULL.
+ * @return A new string which is a duplicate of the input string, or NULL if the input string was NULL.
+ */
+RZ_API char *rz_str_dup(RZ_NULLABLE const char *str) {
```
Comment (wargio) [inline librz/util/str.c]:
```suggestion
RZ_API RZ_OWN char *rz_str_dup(RZ_NULLABLE const char *str) {
```

File: librz/util/str.c [hunk 8]
```diff
@@ -1040,8 +1040,16 @@ RZ_API char *rz_str_ndup(RZ_NULLABLE const char *ptr, int len) {
...
-// TODO: deprecate?
-RZ_API char *rz_str_dup(const char *str) {
+/**
+ * @brief Duplicates a string.
+ *
+ * This function duplicates the given string. If the input string is NULL,
+ * the function will return NULL.
+ *
+ * @param str The string to duplicate. Can be NULL.
+ * @return A new string which is a duplicate of the input string, or NULL if the input string was NULL.
```
Comment (wargio) [inline librz/util/str.c]:
```suggestion
 * \brief Duplicates a given string.
 *
 * This function duplicates the given string. If the input string is NULL, the function will return NULL.
 *
 * \param str The string to duplicate. Can be NULL.
 * \return A new string which is a duplicate of the input string, or NULL if the input string was NULL.
```

**Review by wargio — COMMENTED** [review]
Good job! fix the declarations and it can be merged.

**Comment by XVilka** [comment]
1. `../librz/core/cmd/cmd_eval.c:100: error: identifier expected`
2. Please rebase on top of the latest `dev`

File: librz/core/tui/panels.c (hunk 5 above)
Comment (HN026) [reply librz/core/tui/panels.c]: Okayy

File: librz/core/cmd/cmd_eval.c [hunk 9]
```diff
@@ -95,7 +97,9 @@ RZ_API bool rz_core_theme_load(RzCore *core, const char *name) {
 	goto fail;
 
 success:
-	core->curtheme = rz_str_dup(core->curtheme, name);
+	char *tmp = rz_str_dup(name);
+	free(core->curtheme);
+	core->curtheme = tmp;
```
Comment (wargio) [inline librz/core/cmd/cmd_eval.c]:
```suggestion
	{
		char *tmp = rz_str_dup(name);
		free(core->curtheme);
		core->curtheme = tmp;
	}
```

**Review by XVilka — CHANGES_REQUESTED** [review]
```
Run python3 rz-bindgen/src/lint.py \
Translation unit diagnostic at <rizin/librz/core/cmd/cmd_eval.c:100:2>: expected expression
Translation unit diagnostic at <rizin/librz/core/cmd/cmd_eval.c:102:19>: use of undeclared identifier 'tmp'
Process completed with exit code 2.
```
Also formatting.

**Comment by HN026** [comment]
> 1. `../librz/core/cmd/cmd_eval.c:100: error: identifier expected`
> 
>     2. Please rebase on top of the latest `dev`


> ```
> Run python3 rz-bindgen/src/lint.py \
> Translation unit diagnostic at <rizin/librz/core/cmd/cmd_eval.c:100:2>: expected expression
> Translation unit diagnostic at <rizin/librz/core/cmd/cmd_eval.c:102:19>: use of undeclared identifier 'tmp'
> Process completed with exit code 2.
> ```
> 
> Also formatting.

I've formatted all the files I've touched, Idk why it's breaking there.

**Comment by XVilka** [comment]
See the CI logs for the error messages, formatting also.

**Comment by HN026** [comment]
> See the CI logs for the error messages, formatting also.

Sure

**Comment by Rot127** [comment]
@HN026 The error might come from a different clang-format version on your side. Currently we have clang-format-16 in the CI.

**Comment by HN026** [comment]
> @HN026 The error might come from a different clang-format version on your side. Currently we have clang-format-16 in the CI.

Yup I noticed will be pushing changes by evening.

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**